### DISCUSSION TITLE
External port expander module

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/lizard#34] [Ideas]
To increase the number of available pins and to simplify wiring, we've been thinking about creating a bus system for external port expanders. Here is a short sketch of our current idea:

- [x] The existing `Expander` module remains unchanged.
- [x] We introduce a `core.ota(wifi_ssid, wifi_password, ota_url)` method for firmware updates "over the air".
- [x] After an OTA the ESP calls the server again to verify that an OTA would work with the new firmware. Otherwise the ESP rolls back.
- [ ] We introduce a new `ExternalExpander(serial, address)` module for communicating with _external_ port expanders, which is very similar to the existing `Expander` module.
- [ ] When flashing the external ESP, we call a new `core.store(key, value)` command to write its address (key `__address`) to the NVS.
- [ ] We introduce an `ExternalExpander.ota(...)` method which calls `core.ota(...)` on the external ESP using its address (see syntax below).

Usage example:
```py
serial = Serial(26, 27, 11500, 1)
expander = Expander(serial, 32, 33)
led = expander.Output(15)
led.on()

serial2 = Serial(42, 43, 11500, 2)
external = ExternalExpander(serial2, 0x3)
led = external.Output(15)
led.on()
```

Syntax for inter-ESP communication with leading two-digit hex address
```py
!00core.print(42)@f0
!01core.print(1 + 1)@87
...
!0Fcore.print("foo")@09
```

**Comment by rodja** [comment]
We should try to make the address-schema and baud rates compatible with modbus.

**Reply by Johannes-Thiel** [reply]
Modbus messages will start with a colon : with this difference, there should be no interference with our proposed protocol.

**Comment by JensOgorek** [comment]
I created a pull request (#44) containing the OTA functionality via HTTPS OTA update. It is controlled by the core module with ```core.ota(wifi_ssid, wifi_password, ota_url) ```. There is also an example of an HTTP server receiving the request and providing the binary.

**Comment by falkoschindler** [comment]
See PR #64 "External Expander".

**Comment by falkoschindler** [comment]
I'm closing this discussion because it's mostly implemented using the new Serial Bus with OTB (over-the-bus) updates.
